File: examples/sidecar-js/decision-audit-reporter/src/audit-report-builder.ts
import { CoverageAnalyzer } from './coverage-analyzer.js';
import type {
  AuditReport,
  BoundaryTestResult,
  CoverageResult,
  FniDetails,
} from './types.js';

export const DEFAULT_SLA_THRESHOLD_US = 100_000;

export interface AuditReportBuildInput {
  collectionWindowMinutes: number;
  fniDetails: FniDetails[];
  boundaryResults: Map<string, BoundaryTestResult[]>;
  allRuleIdsByModel: Map<string, string[]>;
  slaThresholdUs?: number;
}

export class AuditReportBuilder {
  private static readonly coverageAnalyzer = new CoverageAnalyzer();

  static build(input: AuditReportBuildInput): AuditReport {
    const slaThresholdUs = input.slaThresholdUs ?? DEFAULT_SLA_THRESHOLD_US;
    const groupedByDecisionRef = groupFniDetailsByDecisionRef(input.fniDetails);

    const decisionRefs = new Set<string>([
      ...groupedByDecisionRef.keys(),
      ...input.boundaryResults.keys(),
      ...input.allRuleIdsByModel.keys(),
    ]);

    const perModel = [...decisionRefs].sort().map((decisionRef) => {
      const modelFnis = groupedByDecisionRef.get(decisionRef) ?? [];
      const latencies = modelFnis.map((fni) => fni.durationUs);
      const ruleIds = input.allRuleIdsByModel.get(decisionRef) ?? [];
      const ruleCoverage = AuditReportBuilder.coverageAnalyzer.analyze(modelFnis, ruleIds);

      return {
        decisionRef,
        executionCount: modelFnis.length,
        avgLatencyUs: average(latencies),
        ruleCoverage,
        boundaryTestResults: input.boundaryResults.get(decisionRef) ?? [],
      };
    });

    const allLatencies = input.fniDetails.map((fni) => fni.durationUs);
    const boundaryTests = [...input.boundaryResults.values()].flat();
    const boundaryErrorCount = boundaryTests.filter((test) => test.error !== null).length;

    return {
      generatedAt: new Date().toISOString(),
      collectionWindowMinutes: input.collectionWindowMinutes,
      summary: {
        totalDecisionExecutions: input.fniDetails.length,
        uniqueDecisionModels: decisionRefs.size,
        avgLatencyUs: average(allLatencies),
        p95LatencyUs: percentile(allLatencies, 95),
        errorRate:
          boundaryTests.length > 0 ? boundaryErrorCount / boundaryTests.length : 0,
      },
      perModel,
      compliance: {
        allModelsEvaluated: perModel.every((model) => model.executionCount > 0),
        noDeadRulesFound: perModel.every((model) => model.ruleCoverage.deadRules.length === 0),
        latencyWithinSla: perModel.every((model) => model.avgLatencyUs < slaThresholdUs),
      },
    };
  }
}

function groupFniDetailsByDecisionRef(
  fniDetails: FniDetails[],
): Map<string, FniDetails[]> {
  const grouped = new Map<string, FniDetails[]>();
  for (const fni of fniDetails) {
    const existing = grouped.get(fni.decisionRef) ?? [];
    existing.push(fni);
    grouped.set(fni.decisionRef, existing);
  }
  return grouped;
}

function average(values: number[]): number {
  if (values.length === 0) return 0;
  return Math.round(values.reduce((sum, value) => sum + value, 0) / values.length);
}

function percentile(values: number[], percentileRank: number): number {
  if (values.length === 0) return 0;
  const sorted = [...values].sort((left, right) => left - right);
  const index = Math.ceil((percentileRank / 100) * sorted.length) - 1;
  return sorted[Math.max(0, Math.min(index, sorted.length - 1))];
}

File: examples/plugins/business_rules/decision_audit_reporter/lib/audit_report_builder.ex
defmodule Examples.BusinessRules.DecisionAuditReporter.AuditReportBuilder do
  @moduledoc """
  Assembles a post-execution decision audit report with per-model coverage,
  boundary-test results, and compliance flags for operators and DMN observers.
  """

  alias Examples.BusinessRules.DecisionAuditReporter.CoverageAnalyzer

  @default_sla_threshold_us 100_000

  @doc "Default SLA threshold in microseconds (100 ms)."
  @spec default_sla_threshold_us() :: pos_integer()
  def default_sla_threshold_us, do: @default_sla_threshold_us

  @doc """
  Builds the audit report.

  `input` keys:

  * `:collection_window_minutes`
  * `:flow_node_instance_details`
  * `:boundary_results` — `%{decision_ref => [boundary_test_result]}`
  * `:all_rule_ids_by_model` — `%{decision_ref => [rule_id]}`
  * `:sla_threshold_us` — optional, default `100_000`
  """
  @spec build(map()) :: map()
  def build(input) when is_map(input) do
    sla_threshold_us = Map.get(input, :sla_threshold_us, @default_sla_threshold_us)
    flow_node_instance_details = Map.get(input, :flow_node_instance_details, [])
    boundary_results = Map.get(input, :boundary_results, %{})
    all_rule_ids_by_model = Map.get(input, :all_rule_ids_by_model, %{})

    grouped = Enum.group_by(flow_node_instance_details, & &1.decision_ref)

    decision_refs =
      [Map.keys(grouped), Map.keys(boundary_results), Map.keys(all_rule_ids_by_model)]
      |> List.flatten()
      |> Enum.uniq()
      |> Enum.sort()

    per_model =
      Enum.map(decision_refs, fn decision_ref ->
        model_instances = Map.get(grouped, decision_ref, [])
        latencies = Enum.map(model_instances, & &1.duration_us)
        rule_ids = Map.get(all_rule_ids_by_model, decision_ref, [])

        %{
          decision_ref: decision_ref,
          execution_count: length(model_instances),
          avg_latency_us: average(latencies),
          rule_coverage: CoverageAnalyzer.analyze(model_instances, rule_ids),
          boundary_test_results: Map.get(boundary_results, decision_ref, [])
        }
      end)

    all_latencies = Enum.map(flow_node_instance_details, & &1.duration_us)
    boundary_tests = boundary_results |> Map.values() |> List.flatten()
    boundary_error_count = Enum.count(boundary_tests, &(&1.error != nil))

    %{
      generated_at: DateTime.utc_now() |> DateTime.to_iso8601(),
      collection_window_minutes: Map.get(input, :collection_window_minutes, 0),
      summary: %{
        total_decision_executions: length(flow_node_instance_details),
        unique_decision_models: length(decision_refs),
        avg_latency_us: average(all_latencies),
        p95_latency_us: percentile(all_latencies, 95),
        error_rate:
          if(boundary_tests == [],
            do: 0,
            else: boundary_error_count / length(boundary_tests)
          )
      },
      per_model: per_model,
      compliance: %{
        all_models_evaluated: Enum.all?(per_model, &(&1.execution_count > 0)),
        no_dead_rules_found: Enum.all?(per_model, &(&1.rule_coverage.dead_rules == [])),
        latency_within_sla: Enum.all?(per_model, &(&1.avg_latency_us < sla_threshold_us))
      }
    }
  end

  defp average([]), do: 0

  defp average(values) do
    round(Enum.sum(values) / length(values))
  end

  defp percentile([], _percentile_rank), do: 0

  defp percentile(values, percentile_rank) do
    sorted_values = Enum.sort(values)
    index = max(0, ceil(length(sorted_values) * percentile_rank / 100) - 1)
    Enum.at(sorted_values, index)
  end
end
